### PR TITLE
[SPARK-25085][SQL] Table subdirectories should inherit the warehouse directory permissions

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Insert overwrite a `non-partitioned` table should not delete table folder because it may contains information like ACL entries.
This pr fix this issue, thus the data source table and the hive table have the same behavior.

This change is the same as [HIVE-3756](https://issues.apache.org/jira/browse/HIVE-3756): Warehouse table subdirectories should inherit the group permissions of the warehouse parent directory.

A simple benchmark:

Number files | Set permission elapsed time(milliseconds)
-- | --
200 | 374
2000 | 3065
10000 | 16334
100000 | 171664

## How was this patch tested?

unit tests and manual tests

manual tests:
```shell
spark-sql -e "create table spark_25085(id int)"
hdfs dfs -getfacl /user/hive/warehouse/spark_25085
hdfs dfs -setfacl -m user:SPARK-25085:rw- /user/hive/warehouse/spark_25085
spark-sql -e "insert overwrite table spark_25085 values(1), (2)"
hdfs dfs -getfacl /user/hive/warehouse/spark_25085
```

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala [hunk 1]
```diff
@@ -124,6 +124,25 @@ abstract class FileCommitProtocol {
...
+   *
+   * Return true if this directory is empty or deleted all files in this directory, otherwise false.
+   */
+  def truncateDirectoryWithJob(fs: FileSystem, directory: Path, recursive: Boolean): Boolean = {
+    assert(fs.isDirectory(directory))
+    val listStatus = fs.listStatus(directory)
+    if (listStatus.isEmpty) {
+      true
+    } else {
+      val result = listStatus.map { f =>
```
Comment (mgaido91) [inline core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala]: what about `forall`?

**Comment by mgaido91** [comment]
I am just worried/concerned by the performance regression this might introduce in the case the table is made of many files. Can we have a benchmark using different file system (at least HDFS, S3) with hundreds and thousands of files? Thanks.

**Comment by wangyum** [comment]
Number files | Delete directory elapsed time(milliseconds) | Truncate directory elapsed   time(milliseconds)
-- | -- | --
200 | 7 | 319
5000 | 48 | 7810
10000 | 90 | 16427
100000 | 736 | 194470

**Comment by wangyum** [comment]
I think about how to improve performance.

**Comment by mgaido91** [comment]
yes, thanks for the benchmark.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -626,6 +626,14 @@ object SQLConf {
     .stringConf
     .createWithDefault("parquet")
 
+  val DATA_SOURCE_TABLE_INHERIT_PERMS = buildConf("spark.sql.datasource.table.inherit.perms")
+    .internal()
+    .doc("Set this to true if the table directories should be inheriting the permission " +
+      "of the warehouse or database directory " +
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I think this is not very clear: it seems to me that the permissions are inherited from the base directory where the table is going to be written. On the other hand, instead, the permissions are inherited from the previous table directory which was deleted, right? Can we make this more clear?

**Review by mgaido91 — COMMENTED** [review]
the approach SGTM.

Have you tried this change on different file systems? Can we ensure this works also with S3, ADLS and not only with HDFS?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 3]
```diff
@@ -261,4 +273,67 @@ case class InsertIntoHadoopFsRelationCommand(
...
+      group: String,
+      permission: FsPermission,
+      aclStatus: AclStatus,
+      fs: FileSystem,
+      target: Path): Unit = {
+    try {
+      // use FsShell to change group, permissions, and extended ACL's recursively
+      val fsShell = new FsShell
+      fsShell.setConf(hadoopConf)
+      fsShell.run(Array[String]("-chgrp", "-R", group, target.toString))
```
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: These code copy from https://github.com/apache/hive/blob/rel/release-1.2.2/shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java#L722-L767

I will think about whether there is a better implementation.

[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: Well, for things like setting the group/user you can use `fs.setOwner` for instance

**Comment by mgaido91** [comment]
cc @cloud-fan @dongjoon-hyun  @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Copy from `hive.warehouse.subdir.inherit.perms`: https://github.com/apache/hive/blob/rel/release-1.2.2/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java#L1729

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: yes, but that does a different thing IIUC, ie. that copies the permissions from the warehouse dir for the new tables created there. Here we are not doing the same, we are copying the permissions from the previous table dir to the new table dir after the overwrite.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 4]
```diff
@@ -257,7 +257,102 @@ class HiveCatalogedDDLSuite extends DDLSuite with TestHiveSingleton with BeforeA
         spark.sql("ALTER TABLE t1 ADD COLUMNS (newcol1 STRUCT<`$col1`:STRING, col2:Int>)")
       }.getMessage
       assert(err.contains("Cannot recognize hive type string:"))
-   }
+    }
+  }
+
+  test("Normal table should inherit database permission") {
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: @wangyum . This inherits the parent directory properties and the property name is `DATA_SOURCE_TABLE_INHERIT_PERMS`. Hence, it's not Hive-catalog only stuff.

It would be great if we can have this test coverage on `InMemoryCatalogedDDLSuite`, not only `HiveCatalogedDDLSuite`. We need some refactoring and the following will be the only difference.
```scala
spark.sql("create table datasource_table(a int) using parquet")
spark.sql("create table hive_table(a int)")
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 5]
```diff
@@ -261,4 +272,69 @@ case class InsertIntoHadoopFsRelationCommand(
...
+      pathExists: Boolean,
+      fs: FileSystem,
+      file: Path): (Option[FsPermission], Option[AclStatus]) = {
+    var permission: Option[FsPermission] = None
+    var aclStatus: Option[AclStatus] = None
+    if (conf.isDataSouceTableInheritPerms && pathExists) {
+      permission = Some(fs.getFileStatus(file).getPermission)
+      if (isExtendedAclEnabled(hadoopConf)) aclStatus = Some(fs.getAclStatus(file))
+    }
+    (permission, aclStatus)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: I would remove `var` by, for instance:

```scala
val shouldFetchPermission = conf.isDataSouceTableInheritPerms && pathExists
val shouldFetchAclStatus = shouldFetchPermission && isExtendedAclEnabled(hadoopConf)

val permission = if (shouldFetchPermission) {
  Some(fs.getFileStatus(file).getPermission)
} else {
  None
}
val aclStatus = if (shouldFetchAclStatus) {
  Some(fs.getAclStatus(file))
} else {
  None
}

(permission, aclStatus)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 6]
```diff
@@ -70,7 +76,6 @@ case class InsertIntoHadoopFsRelationCommand(
     val hadoopConf = sparkSession.sessionState.newHadoopConfWithOptions(options)
     val fs = outputPath.getFileSystem(hadoopConf)
     val qualifiedOutputPath = outputPath.makeQualified(fs.getUri, fs.getWorkingDirectory)
-
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: I would remove unrelated newline changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Hive configuration doc itself doesn't looks quite clear to me as well actually.

**Comment by wangyum** [comment]
Close it because [HIVE-16392](https://jira.apache.org/jira/browse/HIVE-16392): Remove hive.warehouse.subdir.inherit.perms and all permissions inheritance logic.
